# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

首先导入数据分析需要的pandas库，通过pandas库的`read_csv`函数读取原始文件"e_commerce.csv"，将其解析为DataFrame，并赋值给变量`original_data`。查看该DataFrame的前五行数据。

In [1]:
import pandas as pd

In [2]:
original_data=pd.read_csv("e_commerce.csv")

In [3]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

第二步将针对上一步建立的`original_data`这个DataFrame所包含的数据进行评估。  

评估数据将从两个方面进行：结构和内容，即整齐度和干净度。数据结构性问题是指不符合“1.每列是一个变量，2.每行是一个观察值，3.每个单元格是一个值”这三个标准；数据内容性的问题包括存在丢失数据、重复数据、无效数据等情况。

### 评估数据整齐度

In [4]:
original_data.sample(10)

InvoiceNo StockCode                       Description  Quantity  \
324281    565396     23282   FOLDING BUTTERFLY MIRROR IVORY          1   
339489    566597     22699  ROSES REGENCY TEACUP AND SAUCER          6   
314576    564646     22079        RIBBON REEL HEARTS DESIGN          2   
29391     538783     21891  TRADITIONAL WOODEN SKIPPING ROPE        12   
527024    580690     84819      DANISH ROSE ROUND SEWING BOX        32   
171265    551399     21622  VINTAGE UNION JACK CUSHION COVER         8   
386108    570225     16238         PARTY TIME PENCIL ERASERS         5   
431191    573585     48188           DOORMAT WELCOME PUPPIES         2   
84169     543370     22207             FRYING PAN UNION FLAG         4   
537157    581238     22630              DOLLY GIRL LUNCH BOX         2   

             InvoiceDate  UnitPrice  CustomerID         Country  
324281    9/2/2011 16:39       2.46         NaN  United Kingdom  
339489   9/13/2011 15:23       2.95     14194.0  United Kingdom  
314576   8/26/2011 13:39       2.46         NaN  United Kingdom  
29391   12/14/2010 11:38       1.25     14961.0  United Kingdom  
527024   12/5/2011 15:44       0.79     16597.0  United Kingdom  
171265   4/28/2011 13:17       4.95     17874.0  United Kingdom  
386108   10/9/2011 13:15       0.21     16595.0  United Kingdom  
431191  10/31/2011 14:41      15.79         NaN  United Kingdom  
84169     2/7/2011 14:51       4.25     12359.0          Cyprus  
537157   12/8/2011 10:53       2.46         NaN  United Kingdom

抽取10条样本观察，数据符合“1.每列是一个变量，2.每行是一个观察值，3.每个单元格是一个值”的标准，具体来看，每行是一次具体交易，每列是关于交易的一个变量，单元格内是关于该交易该变量的值，因此不存在结构性问题。

### 评估数据干净度

In [5]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果来看，数据共有541909条观察值，其中`Description`和`CustomerID`存在缺失值；`InvoiceDate`应该为日期类型，`CustomerID`应该是字符串类型，需要进行格式转换。

#### 评估缺失数据

在了解`Description`存在缺失值后，根据条件提出缺失观察值。

In [6]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

有1454条交易记录缺失`Dscription`变量值。  
从输出结果来看，`Dscription`缺失的值对应的`UnitPrice`均为0。为验证猜想，增加筛选条件，看是否存在`Dscription`数据缺失但`UnitPrice`不为0的数据。

In [7]:
original_data[(original_data["Description"].isnull())&(original_data["UnitPrice"]!=0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

输出结果数量为0条，说明不存在`Dscription`缺失的数据，同时不具备有效的`UnitPrice`值。  

`Dscription`表示产品名称，`UnitPrice`表示产品单价，都是进行后续交易分析的重要变量。如果他们同时缺失或者无效，那么认为该数据无法提供有效含义，因此后续可以进行删除。

`CustomerID`也存在缺失值，因此同样根据条件提取出观察缺失值。

In [8]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

有135080条数据缺失`CustomerID`值。`CustomerID`表示客户编号，不是分析畅销产品的关键数据，从输出结果看，一些`CustomerID`值缺失的数据依旧有效，因此保留变量为空的观察值。

#### 评估重复数据

根据数据变量的含义来看，`InvoiceNo`、`StockCode`和`CustomerID`都是唯一标识，但一次交易可能包含多件商品，因此`InvoiceNo`可以存在重复；同一件商品可能存在于不同的交易中，因此`StockCode`也可以存在重复；同一客户可能进行多次交易，因此`CustomerID`也可以存在重复。  

那么针对此数据集，无需评估重复数据。

#### 评估不一致数据

不一致数据可能存在于`Country`变量中，需要查看是否存在多个不同值代指一个国家的情况。

In [9]:
original_data["Country"].value_counts().sort_index()

Australia                 1259
Austria                    401
Bahrain                     19
Belgium                   2069
Brazil                      32
Canada                     151
Channel Islands            758
China                      288
Cyprus                     622
Czech Republic              30
Denmark                    389
EIRE                      8196
European Community          61
Finland                    695
France                    8557
Germany                   9495
Greece                     146
Iceland                    182
Israel                     297
Italy                      803
Japan                      358
Lebanon                     45
Lithuania                   35
Malta                      127
Netherlands               2371
Norway                    1086
Poland                     341
Portugal                  1519
RSA                         58
Saudi Arabia                10
Singapore                  229
Spain                     2533
Sweden  

从`Country`变量来看，"USA"、"United States"都指代美国，"UK""U.K.""United Kindom"都指代英国，因此在清洗步骤中需要对这些值进行统一，只保留一个指代值。

#### 评估无效数据或错误数据

可以通过DataFrame的`describe`方法对数值的统计信息进行了解。

In [10]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从输出结果来看，`Quantity`和`UnitPrice`存在负数，会对后续分析产生影响。  

因此我们先筛选出`Quantity`为负数的观察值，进一步评估其含义。

In [11]:
original_data[original_data["Quantity"]<0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

从筛选结果来看，`Quantity`变量为负数的观察值，`InvoiceNo`似乎均以C开头，表示订单被取消。

但此猜测不一定准确，为了验证猜想，增加筛选条件，看是否存在`Quantity`变量为负数且`InvoiceNo`不以C开头的观察值。

In [12]:
original_data[(original_data["Quantity"]<0)&(original_data["InvoiceNo"].str[0]!="C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

以上猜想被证明错误，因为还存在`Quantity`变量为负数且`InvoiceNo`不以C开头的观察值。  

但根据以上输出结果，筛选出的观察值`UnitPrice`观察值均为0，因此增加条件验证`UnitPrice`。

In [13]:
original_data[(original_data["Quantity"]<0)&(original_data["InvoiceNo"].str[0]!="C")&(original_data["UnitPrice"]!=0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

输出结果数量为0条，说明不存在`Quantity`变量为负数、`InvoiceNo`不以C开头的并且`UnitPrice`不为0的值。  

根据以上结果，猜想得到验证，当`Quantity`变量为负数时，观察值满足以下条件之一：  
 - `InvoiceNo`以"C"开头，表示订单被取消  
 - `UnitPrice`为0，表示单价为0英镑  
 
订单被取消和单价为0的数据都是不是有效成交数据，不贡献销售，不在后续分析范围内，因此在数据清理步骤中，可以将`Quantity`为负数的观察值删除。

接下来筛选`UnitPrice`为负数的观察值。

In [14]:
original_data[original_data["UnitPrice"]<0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

从输出结果来看，`UnitPrice`为负数的观察值都是坏账调整，不属于商品交易的数据，因此在清洗步骤中也将其删除。

## 清理数据

第三步将根据评估的结果进行数据清理。需要进行的数据清理包括： 

- 把`InvoiceDate`变量的数据类型转换为日期时间类型
- 把`CustomerId`变量的数据类型转换为字符串类型
- 把`Description`变量缺失的观察值删除
- 把`Country`变量值`"USA"`替换为`"United States"`
- 把`Country`变量值`"UK","U.K."`替换为`"United Kindom"`
- 把`Quantity`变量为负数的观察值删除
- 把`UnitPrice`变量为负数的观察值删除  

为了区分经过清理的数据和原始数据，创建新的变量`cleaned_data`，让它成为`original_data`的副本。之后的清理步骤将被运用在`cleaned_data`上。

In [15]:
cleaned_data=original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

#### 把`InvoiceDate`变量的数据类型转换为日期时间类型：

In [16]:
cleaned_data["InvoiceDate"]=pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把`CustomerId`变量的数据类型转换为字符串类型：

In [17]:
cleaned_data["CustomerID"]=cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

把`CustomerId`变量结尾的`".0"`删除：

In [20]:
cleaned_data["CustomerID"]=cleaned_data["CustomerID"].str.slice(0,-2)
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把`Description`变量缺失的观察值删除，并检查删除后该列空缺值个数和：

In [21]:
cleaned_data.dropna(subset=["Description"],inplace=True)

In [23]:
cleaned_data["Description"].isnull().sum()

0

把`Country`变量值`"USA"`替换为`"United States"`,并检查替换后`"USA"`变量值的个数：

In [26]:
cleaned_data["Country"]=cleaned_data["Country"].replace("USA","United States")

In [27]:
len(cleaned_data[cleaned_data["Country"]=="USA"])

0

把`Country`变量值`"UK"`,`"U.K."`替换为`"United Kindom"`，并检查替换后`"UK"`,`"U.K."`变量值的个数：

In [32]:
cleaned_data["Country"]=cleaned_data["Country"].replace({"UK":"United Kindom","U.K.":"United Kindom"})

In [33]:
print(len(cleaned_data[cleaned_data["Country"]=="UK"]))
print(len(cleaned_data[cleaned_data["Country"]=="U.K."]))

0
0


把`Quantity`变量为负数的观察值删除,并检查删除后该列为负数的个数：

In [36]:
cleaned_data=cleaned_data[cleaned_data["Quantity"]>=0]

In [37]:
len(cleaned_data[cleaned_data["Quantity"]<0])

0

把`UnitPrice`变量为负数的观察值删除,并检查删除后该列为负数的个数：

In [38]:
cleaned_data=cleaned_data[cleaned_data["UnitPrice"]>=0]

In [39]:
len(cleaned_data[cleaned_data["UnitPrice"]<0])

0

## 保存清理后的数据

完成数据清理后，把干净整齐的数据保存到新的文件里，文件名为e_commerce_cleaned.csv。

In [40]:
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom

In [41]:
cleaned_data.to_csv("e_commerce_cleaned.csv",index=False)

In [43]:
pd.read_csv("e_commerce_cleaned.csv").head()

C:\Users\iron\AppData\Roaming\Python\Python39\site-packages\IPython\core\interactiveshell.py:3505: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom

In [45]:
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 530691 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    530691 non-null  object        
 1   StockCode    530691 non-null  object        
 2   Description  530691 non-null  object        
 3   Quantity     530691 non-null  int64         
 4   InvoiceDate  530691 non-null  datetime64[ns]
 5   UnitPrice    530691 non-null  float64       
 6   CustomerID   530691 non-null  object        
 7   Country      530691 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(5)
memory usage: 36.4+ MB
